src/inferenceql/query/plan.cljc

```diff
@@ -335,16 +342,42 @@
     [[:scalar-expr & _]] true
     :else false))
 
+(defn ^:private select-star-plan
+  "Returns a plan for SELECT * and SELECT * EXCEPT ... clauses."
+  [node op]
+  (tree/match [node]
+    [[:select-star-clause & children]]
+    (select-star-plan children op)                          ; trying recur breaks with CLJ-2808
+
+    ;; plain SELECT *
+    [[[:star _]]]
+    op
+
+    ;; SELECT * EXCEPT (col1, col2, ...)
+    [[[:star & _] [:select-except-clause _except _lparen [:identifier-list & id-list] _rparen]]]
+    (loop [exclusions []
+           [id & rest-ids] (filterv (tree/tag-pred :identifier) id-list)]
+      (if id
+        (recur (conj exclusions (literal/read id)) rest-ids)
+        {::type :inferenceql.query.plan.type/star-except
+         ::exclusions exclusions
+         ::plan op}))
+
+    :else
+    (throw (ex-info "Can't generate selection plan - unknown * pattern"
+                    {:star-node node
+                     :partial-plan op}))))
+
 (defn select-plan
   [select-node group-by-node op]
   (assert (= :select-clause (tree/tag select-node)))
   (assert (or (nil? group-by-node) (= :group-by-clause (tree/tag group-by-node))))
-  (let [selections (selections select-node)
+  (let [selections (vec (selections select-node))
         distinct-clause (tree/get-node select-node :distinct-clause)
-        plan (cond (and (= 1 (count selections))
-                        (star? (first selections)))
-                   op
+        plan (cond (star? (first selections))
+                   (select-star-plan (first selections) op)
```

Comment:
can this ever fail? Can we get an empty selections input?

Reply:
Good question. I checked, but all my attempts result in parse errors. I don't think we can reach this code without _something_ in the selections.